### PR TITLE
[FIRRTL] yosys equiv checking infrastructure and FIRRTL primatives testcase

### PR DESCRIPTION
**Pull request by darthscsi**

### CODE REVIEW

**Comment by darthscsi** [comment]
@lattner, would scripts for tests go into tools?

File: CMakeLists.txt [hunk 1]
```diff
@@ -121,6 +121,14 @@ else()
   endif()
 endif()
 
+# If Yosys hasn't been explicitly disabled, find it.
+option(YOSYS_DISABLE "Disable the yosys tests.")
+if (YOSYS_DISABLE)
+  message(STATUS "Disabling yosys tests.")
+else()
+  find_program(YOSYS_PATH "yosys")
+endif()
```
Comment (teqdruid) [inline CMakeLists.txt:130]: This should print out whether or not yosys is has successfully been located.

File: equiv_yosys.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+
+#echo "Comparing $1 and $2 with $3"
+yosys -q -p "
```
Comment (teqdruid) [inline equiv_yosys.sh:4]: Why not run yosys directly on a tcl script? Lit is perfectly capable of detecting exit code failures.

File: test/firtool/verilog_equiv.fir [hunk 3]
```diff
@@ -0,0 +1,245 @@
+; REQUIRES: yosys
```
Comment (teqdruid) [inline test/firtool/verilog_equiv.fir:1]: This seems like more of an integration test to me.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 4]
```diff
@@ -873,7 +873,7 @@ LogicalResult FIRRTLLowering::visitExpr(CvtPrimOp op) {
 
   // Signed to signed is a noop.
   if (getTypeOf<IntType>(op.getOperand()).isSigned())
-    setLowering(op, operand);
+    return setLowering(op, operand);
```
Comment (lattner) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:876]: Please split this out to its own patch w testcase

File: lib/EmitVerilog/EmitVerilog.cpp [hunk 5]
```diff
@@ -739,32 +739,54 @@ class ExprEmitter : public ExprVisitor<ExprEmitter, SubExprInfo>,
   SubExprInfo emitBitSelect(Value operand, unsigned hiBit, unsigned loBit);
 
   SubExprInfo emitBinary(Operation *op, VerilogPrecedence prec,
-                         const char *syntax, bool hasStrictSign = false,
+                         const char *syntax,
```
Comment (lattner) [inline lib/EmitVerilog/EmitVerilog.cpp:742]: I think it would be best to split the yosis integration work out from the behavior changes here.  These are logically distinct changes.

File: lib/EmitVerilog/EmitVerilog.cpp [hunk 6]
```diff
@@ -957,17 +1003,33 @@ SubExprInfo ExprEmitter::emitBinary(Operation *op, VerilogPrecedence prec,
...
-      emitSubExpr(op->getOperand(1), rhsPrec, hasStrictSign, opForceSign);
+  if (op1Signed || opForceSign)
+    os << "$signed(";
+  if (op1Width < width) {
+    if (op1Signed)
+        os << "{{" << (width - op1Width) << '{';
+    else
+        os << "{{" << (width - op1Width) << "'d0}, ";
+  }
+  emitSubExpr(op->getOperand(1), (op1Width >= width) ? rhsPrec : LowestPrecedence, opForceSign);
```
Comment (lattner) [inline lib/EmitVerilog/EmitVerilog.cpp:1014]: Clang format.

File: test/equiv_yosys.sh [hunk 7]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
```
Comment (lattner) [inline test/equiv_yosys.sh:1]: I think it makes sense to put this under circt/utils.  Please add the standard file header boilerplate.  I also think it would make sense to name this `equiv-yosis.sh` (LLVM prefers hyphens instead of underscores in tool names) or perhaps just `equiv-check` since yosis is an implementation detail of the script.

File: test/firtool/verilog_equiv.fir (hunk 3 above)
Comment (lattner) [reply test/firtool/verilog_equiv.fir:1]: Agreed, I think that all of these will be.  I think that having yosis equivalence checking will be a valuable thing, but we should use it very sparingly, and for larger scale integration test sorts of things.

File: test/firtool/verilog_equiv.fir [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+; RUN: split-file %s %t
+; RUN: firtool %t/test_mod.fir --format=fir -verilog -o %t/test_mod.fir.v
+; RUN: firtool %t/test_mod.fir --format=fir -mlir -o %t/test_mod.fir.mlir
+; RUN: firtool %t/test_mod.fir.mlir --lower-to-rtl -o %t/test_mod.rtl.mlir
+; RUN: firtool %t/test_unary.fir --format=fir -verilog -o %t/test_unary.fir.v
+; RUN: firtool %t/test_unary.fir --format=fir -mlir -o %t/test_unary.fir.mlir
+; RUN: firtool %t/test_prim.fir --format=fir -verilog -o %t/test_prim.fir.v
+; RUN: firtool %t/test_prim.fir --format=fir -mlir -o %t/test_prim.fir.mlir
```
Comment (lattner) [inline test/firtool/verilog_equiv.fir:9]: This looks like a test for the EmitVerilog changes above.  Please split these all out from the yosis integration patch and add a couple unit tests for the yosis driver under test/equiv-yosis/ or something.  Those should be written as .v files and shouldn't depend on the other circt tools.

The test for the EmitVerilog changes should be written in the style of unit tests like the other emitverilog tests.

File: test/firtool/verilog_equiv.fir (hunk 3 above)
Comment (darthscsi) [reply test/firtool/verilog_equiv.fir:1]: Larger test cases will be valuable, but the coverage of firrtl combinatorial ops found ~8 ops with incorrect codegen and 1 bug in scala firrtl.  These small tests are also very fast (1 second range).

File: test/firtool/verilog_equiv.fir (hunk 8 above)
Comment (darthscsi) [reply test/firtool/verilog_equiv.fir:9]: The mlir lines are residual of adding lower-to-rtl to verilog checks too.  Those fail right now, though I'll add the RUN lines for them.

File: test/firtool/verilog_equiv.fir (hunk 3 above)
Comment (lattner) [reply test/firtool/verilog_equiv.fir:1]: > Larger test cases will be valuable, but the coverage of firrtl combinatorial ops found ~8 ops with incorrect codegen and 1 bug in scala firrtl. These small tests are also very fast (1 second range).

Sure, but the EmitVerilog tests run in 10s of milliseconds :)

What I'm suggesting is that we split the "find the bugs" phase of this to be an process that is offline from test flows, and keep "verify the bug fixes" online.  The offline process can be really slow, but run against huge corpuses of designs, and be partially manual.  The online tests need to be extremely fast because we'll be running them (and 10's of thousands of their future friends) all the time.

Comment (lattner) [reply test/firtool/verilog_equiv.fir:1]: Also, the yosis tests will only run for people that have yosis installed, we want the regression tests to be run all the time.

**Review by lattner — APPROVED** [review]
Looks great thank you.  Let's not add a bunch of integration tests of this style in the future, but having this one LGTM!  Thanks Andrew

File: test/firtool/verilog_equiv.fir (hunk 3 above)
Comment (teqdruid) [reply test/firtool/verilog_equiv.fir]: The verilator tests in the regression test directory (which check for basic verilator-only functionality) won't run unless one has verilator installed. So there is precedent for that behavior. They, however, run in 80ms each.
